### PR TITLE
[mod] in the preference page, show !bang of subgrouping categories

### PR DESCRIPTION
**Pull request by return42**
The names of the subgrouping categories in the preference page are translated,
to use this categories the user needs to know by which !bang the category can be
selected.  Related to "Make 'non tab category' bangs discoverable" in [#690].

Related:

- [#690] https://github.com/searxng/searxng/issues/690
- https://github.com/searxng/searxng/issues/1604
- https://github.com/searxng/searxng/pull/1545

------

- the bang column has been moved to the right (screenshot header col `!bang`)
- the bang of the subgroup has been added in the separator line (screenshot `!scientific_publications`
- in the tool-tip all !bangs of the engine entry are shown (screenshot tooltip in the bottom right)

grafik

grafik

### CODE REVIEW

**Comment by dalf** [comment]
From my point of view:
* The shortcuts are some alternative names. When bangs are on the right side, and engine names are on the left side, the users have to lock back and forth between the left and right sides of the screen.
* In the tooltip over the engine names: the categories are confusing because `!general` selects more than this engine. I guess that the purpose is to display the bang `!science` which is currently hidden.

Here an alternative layout:
* The bang column expands over multiple columns for the engine group.
* Engine tooltip: the bangs for the categories are displayed on a second line. Labels are on right side: maybe there is a better way.

Note slightly out of topic: The engine name and bangs should share the same tooltip.

image


<details>

<summary>patch to implement this change</summary>

```patch
diff --git a/searx/static/themes/simple/src/less/preferences.less b/searx/static/themes/simple/src/less/preferences.less
index 0f9d7480a..90aa0b492 100644
--- a/searx/static/themes/simple/src/less/preferences.less
+++ b/searx/static/themes/simple/src/less/preferences.less
@@ -73,7 +73,7 @@
   }
 
   .bang {
-    .ltr-float-right();
+    text-align: left;
     .rounded-corners-tiny;
     background-color: var(--color-doc-code-background);
     color: var(--color-doc-code);
diff --git a/searx/templates/simple/preferences.html b/searx/templates/simple/preferences.html
index dc0bccf14..41646e744 100644
--- a/searx/templates/simple/preferences.html
+++ b/searx/templates/simple/preferences.html
@@ -32,7 +32,8 @@
       {{ _('View error logs and submit a bug report') -}}
     </a>
     {%- endif -%}
-    <p>{% for bang in search_engine.categories + [search_engine.name] + [search_engine.shortcut] %}<span class="bang"> {{ '!' + bang.replace(' ', '_') }}</span>{% endfor %}</p>
+    <p><span class="right">{{ _("Bangs for this engine") }}</span>{% for bang in [search_engine.name] + [search_engine.shortcut] %}<span class="bang"> {{ '!' + bang.replace(' ', '_') }}</span>{% endfor %}</p>
+    <p><span class="right">{{ _("Bang(s) for its category") }}</span>{% for bang in search_engine.categories %}<span class="bang"> {{ '!' + bang.replace(' ', '_') }}</span>{% endfor %}</p>
 </div>
 {%- endif -%}
 {%- endmacro %}
@@ -309,17 +310,20 @@
       <tr>{{- "" -}}
         <th class="engine_checkbox">{{ _("Allow") }}</th>{{- "" -}}
         <th class="name">{{ _("Engine name") }}</th>{{- "" -}}
+        <th class="shortcut">{{ _("Bang") }}</th>{{- "" -}}
         <th>{{ _("Supports selected language") }}</th>{{- "" -}}
         <th>{{ _("SafeSearch") }}</th>{{- "" -}}
         <th>{{ _("Time range") }}</th>{{- "" -}}
         {%- if enable_metrics %}<th>{{ _("Response time") }}</th>{% endif -%}
         <th>{{ _("Max time") }}</th>{{- "" -}}
         {%- if enable_metrics %}<th>{{ _("Reliability") }}</th>{% endif -%}
-        <th class="shortcut"><span class="bang">{{ _("!bang") }}</span></th>{{- "" -}}
       </tr>
       {% for group, group_bang, engines in engines_by_category[categ] | group_engines_in_tab %}
       {% if loop.length > 1 %}
-      <tr><th colspan="9" class="engine-group">{{_(group)}}{% if group_bang %}<span class="bang">{{group_bang}}</span>{% endif %}</th></tr>{{- "" -}}
+      <tr>
+        <th class="engine-group" colspan="2">{{_(group)}}</th>
+        <th class="engine-group" colspan="7">{% if group_bang %}<span class="bang">{{group_bang}}</span>{% endif %}</th>
+      </tr>{{- "" -}}
       {% endif %}
       {% for search_engine in engines %}
       {% if not search_engine.private %}
@@ -335,13 +339,13 @@
           </label>
           {{- engine_about(search_engine) -}}
         </th>{{- "" -}}
+        <td class="shortcut"><span class="bang">{{ '!' + shortcuts[search_engine.name] }}</span></td>{{- "" -}}
         <td>{{ checkbox(None, supports[search_engine.name]['supports_selected_language'], true) }}</td>{{- "" -}}
         <td>{{ checkbox(None, supports[search_engine.name]['safesearch'], true) }}</td>{{- "" -}}
         <td>{{ checkbox(None, supports[search_engine.name]['time_range_support'], true) }}</td>{{- "" -}}
         {%- if enable_metrics %}{{- engine_time(search_engine.name) -}}{% endif -%}
         <td class="{{ 'danger' if stats[search_engine.name]['warn_timeout'] else '' }}">{{ search_engine.timeout }}</td>{{- "" -}}
         {%- if enable_metrics %}{{ engine_reliability(search_engine.name) -}}{% endif -%}
-        <td class="shortcut"><span class="bang">{{ '!' + shortcuts[search_engine.name] }}</span></td>{{- "" -}}
       </tr>
       {% endif %}
       {% endfor %}
```

</details>

**Comment by return42** [comment]
> Here an alternative layout:

can you push your commit on this branch / thanks!

**Comment by dalf** [comment]
Done without the `static.build.commit`

**Comment by dalf** [comment]
to fix : `Bang(s) for its category` --> `Bang(s) for its categories` (since there are many a category and a subcategory).

**Comment by return42** [comment]
> to fix : `Bang(s) for its category` --> `Bang(s) for its categories` (since there are many a category and a subcategory).

I fixed it to::

```diff
diff --git a/searx/templates/simple/preferences.html b/searx/templates/simple/preferences.html
index 41646e744..170dffa39 100644
--- a/searx/templates/simple/preferences.html
+++ b/searx/templates/simple/preferences.html
@@ -32,8 +32,8 @@
       {{ _('View error logs and submit a bug report') -}}
     </a>
     {%- endif -%}
-    <p><span class="right">{{ _("Bangs for this engine") }}</span>{% for bang in [search_engine.name] + [search_engine.shortcut] %}<span class="bang"> {{ '!' + bang.replace(' ', '_') }}</span>{% endfor %}</p>
-    <p><span class="right">{{ _("Bang(s) for its category") }}</span>{% for bang in search_engine.categories %}<span class="bang"> {{ '!' + bang.replace(' ', '_') }}</span>{% endfor %}</p>
+    <p><span class="right">{{ _("!bang for this engine") }}</span>{% for bang in [search_engine.name] + [search_engine.shortcut] %}<span class="bang"> {{ '!' + bang.replace(' ', '_') }}</span>{% endfor %}</p>
+    <p><span class="right">{{ _("!bang for its categories") }}</span>{% for bang in search_engine.categories %}<span class="bang"> {{ '!' + bang.replace(' ', '_') }}</span>{% endfor %}</p>
 </div>
 {%- endif -%}
 {%- endmacro %}
```

OK?

File: searx/static/themes/simple/src/less/preferences.less [hunk 1]
```diff
@@ -72,6 +72,15 @@
     font-size: 90%;
   }
 
+  .bang {
+    text-align: left;
```
Comment (dalf) [inline searx/static/themes/simple/src/less/preferences.less]: Ahh.. I forgot: it should be `.ltr-text-align-left()` so it works also with RTL languages.

Comment (return42) [reply searx/static/themes/simple/src/less/preferences.less]: fixed ... if you think PR is OK, press merge button .. / I have to go for now / thanks for your improvements :+1:

**Pull request merged**